install_config/router/customized_haproxy_router.adoc

```diff
@@ -13,48 +13,476 @@ toc::[]
+global
+{{ end }}
+----
+
+The above will copy *global* to the */var/lib/haproxy/conf/haproxy.config* file
+and then close the file.
+
+Lets set up logging based on environment variables.
+
+----
+{{ with (env "ROUTER_SYSLOG_ADDRESS" "") }}
+  log {{.}} {{env "ROUTER_LOG_FACILITY" "local1"}} {{env "ROUTER_LOG_LEVEL" "warning"}}
+{{ end }}
+----
+
+The environment variables are a map with the environment variable name as
+the key and the contents as the value. The *env* function extracts the value
+for the environment variable.  if the environment variable is not defined or
+*nil* the second argument is returned.
+
+The the result of evaluating what is between the "()" goes in "." (dot).
+The *with* action tests Dot for *nil*. If not *nil* the clause
+is processed up to the *end*. In the above assume `ROUTER_SYSLOG_ADDRESS`
+contains "/var/log/msg", `ROUTER_LOG_FACILITY` is not defined and
+`ROUTER_LOG_LEVEL` contains "info". The following will be copied to the
+output file:
+
+----
+  log /var/log/msg local1 info
+----
+
+Each admitted route ends up generating lines in the configuration file.
+Lets go through the admitted routes:
+
+----
+{{ range $cfgIdx, $cfg := .State }}
+  backend be_http_{{$cfgIdx}}
+{{end}}
+----
+
+*.State* is a map of *ServiceAliasConfig* the key is the route name.
+*range* steps through the map and for each pass it sets $cfgIdx with the *key*,
+and sets $cfg to point to the *ServiceAliasConfig* that describes the route. If
+there are two routes named *myroute*, *hisroute*, the above will copy
+
+----
+  backend be_http_myroute
+  backend be_http_hisroute
+----
+
+to the output file.
+
+Route Annotations, *$cfg.Annotations*, is also a map with the annotation
+name as the key and the content string as the value.  The route can have
+as many annotations as desired and the use is defined by the developer.
+The user codes the annotation into the route and the developer customized
+the HAProxy template to handle the annotation.
+
+The common usage is to index the annotation to get the value. (*env* does
+this for environment variables).
+
+----
+{{$balanceAlgo := index $cfg.Annotations "haproxy.router.openshift.io/balance"}}
+----
+The index extracts the value for the given annotation, if any.
+So $balanceAlgo will contain the string associated with the annotation or *nil*.
+As above, you can test for a non *nil* string and act on it with the *with*
+construct.
+
+----
+{{ with $balanceAlgo }}
+  balance $balanceAlgo
+{{ end }}
+----
+Here when $balanceAlgo is not *nil*, "balance $balanceAlgo" is copied to the
+output file.
+
+In a second example we want to set a server timeout based on a timeout value
+ set in an annotation.
+
+----
+$value := index $cfg.Annotations "haproxy.router.openshift.io/timeout"
+----
+The $value can now be evaluated to make sure it contains a properly constructed
+string.  The *matchPattern* function accepts a regular expression and returns
+TRUE if the argument satisfies the expression.
```

Comment:
It might be clearer to write "*true*" instead of "TRUE" to emphasise that this is a Boolean value, not a string.